### DISCUSSION TITLE
The value of 'refresh' should be used.

### DISCUSSION THREAD

**Discussion by shahmirzali49** [rrousselGit/riverpod#1965] [Q&A] [answered]
@rrousselGit  The value of 'refresh' should be used.  I'm using ref.refresh() different page that there isn't using the provider's value.  Is it a problem? for example, I have 2 pages. On the first page, I have a basket page. second basket product detail page. I wanna refresh the basket data from the basket detail page which I didn't use basket data in the basket product detail page. in this scenario, I used ref.refresh() 

<img width="416" alt="Screenshot 2022-12-01 at 16 36 20" src="https://user-images.githubusercontent.com/49102327/205066597-7429a499-5bce-4f30-a873-fdeffe29b931.png">

**Comment by rrousselGit** [comment, marked as answer]
As the documentation of `refresh` states, if you don't use its return value, instead use `invalidate`

**Reply by shahmirzali49** [reply]
thanks for reply

**Reply by msarkrish** [reply]
@rrousselGit but in my application, i have used refresh in many screens. After migrating to latest riverpod version in some scenarios if I use ref.refresh then only futureprovider is calling and application working correctly. But If I change it to ref.invalidate it is not working as expected. 

Like wise if I use ref.invalidate in some screen then it is calling future provider again and working correctly but if I call ref.refresh then it is not working as expected. 

Can you suggest me correct way @rrousselGit 

In my application in many screens i have to refresh other providers based on some condition. But after migrating to latest riverpod version in many screens functionality is not working which is correctly working with old riverpod version.

**Reply by shahmirzali49** [reply]
@msarkrish look at when properties. skipLoadingOnRefresh and skipLoadingOnReload. 

<img width="261" alt="Screenshot 2022-12-01 at 20 52 28" src="https://user-images.githubusercontent.com/49102327/205125079-51e8c40f-585a-44bc-b897-518cbed76305.png">


and also If you want to every time loading in your screen use unwrapPrevious()

`final getBasketViewModel = ref.watch(getBasketViewModelProvider).unwrapPrevious();`

**Reply by rrousselGit** [reply]
@msarkrish `invalidate` _is_ the correct way. If that doesn't work for you, you most likely have a bug in your codebase that you didn't spot before.

**Reply by msarkrish** [reply]
> @msarkrish look at when properties. skipLoadingOnRefresh and skipLoadingOnReload. 
> 
> <img width="261" alt="Screenshot 2022-12-01 at 20 52 28" src="https://user-images.githubusercontent.com/49102327/205125079-51e8c40f-585a-44bc-b897-518cbed76305.png">
> 
> 
> and also If you want to every time loading in your screen use unwrapPrevious()
> 
> `final getBasketViewModel = ref.watch(getBasketViewModelProvider).unwrapPrevious();`
> 

Thanks for the reply, I will try @Shahmirzali-Huseynov

**Reply by msarkrish** [reply]
> @msarkrish `invalidate` _is_ the correct way. If that doesn't work for you, you most likely have a bug in your codebase that you didn't spot before.

Thanks for the reply @rrousselGit I will check my code.

**Reply by shahmirzali49** [reply]
@msarkrish and also I realized now, If you want to every time show loading on the screen you need to use autoDispose in your FutureProvider. am I right @rrousselGit ? because without autoDispose, unwrapPrevious is not enough itself.

**Reply by felipe-xavier** [reply]
dear @rrousselGit, could you please tell me the documentation link for the refresh and invalidate methods? I couldn't find them on the riverpod website.

**Reply by mono0926** [reply]
@felipe-xavier 

- [refresh method - Ref extension - riverpod library - Dart API](https://pub.dev/documentation/riverpod/latest/riverpod/Ref/refresh.html)
- [invalidate method - Ref extension - riverpod library - Dart API](https://pub.dev/documentation/riverpod/latest/riverpod/Ref/invalidate.html)
